File: day4/solution part 2.py
count = 0

def check(passphrases: list[str]) -> bool:
    seen = set()
    for phrase in passphrases:
        sorted_word = ''.join(sorted(phrase))
        if sorted_word in seen:
            return False
        seen.add(sorted_word)
    return True


with open("input.txt", 'r') as file:
    for line in file:
        passphrases = list(line.split())
        if check(passphrases):
            count += 1

print(count)

File: day4/solution part 2.cpp
#include <unordered_set>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>
#include <iostream>
#include <vector>

bool check(const std::vector<std::string>& passphrase){
    std::unordered_set<std::string> temp;
    for (const std::string& phrase : passphrase){
        if (temp.contains(phrase)) return false;
        temp.insert(phrase);
    }
    return true;
}

int main() {
    int count{};

    std::ifstream in("input.txt"); std::string line, phrase;
    while (std::getline(in, line)){
        std::stringstream ss(line); std::vector<std::string> passphrase;
        while (ss >> phrase){
            std::sort(phrase.begin(), phrase.end());
            passphrase.push_back(phrase);
        }
        if (check(passphrase)) count++;
    }
    std::cout << count;
}